File: tests/unit/cose/hash-envelope.test.ts
/**
 * COSE Hash Envelope Tests
 * Test suite for hash envelope creation and validation (labels 258, 259, 260)
 */

import { describe, test, expect, beforeEach, afterEach } from "bun:test";
import { CoseHashAlgorithm } from "../../../src/lib/cose/constants.ts";
import * as fs from "fs";
import * as path from "path";

describe("Hash Envelope Creation", () => {
  const testDir = "./.test-hash-envelope";

  beforeEach(() => {
    if (fs.existsSync(testDir)) {
      fs.rmSync(testDir, { recursive: true });
    }
    fs.mkdirSync(testDir, { recursive: true });
  });

  afterEach(() => {
    if (fs.existsSync(testDir)) {
      fs.rmSync(testDir, { recursive: true });
    }
  });

  test("can create hash envelope from small data", async () => {
    const { createHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const data = new TextEncoder().encode("Hello, hash envelope!");
    const envelope = await createHashEnvelope(data, {
      contentType: "text/plain",
    });

    expect(envelope.payloadHash).toBeInstanceOf(Uint8Array);
    expect(envelope.payloadHash.length).toBe(32); // SHA-256 = 32 bytes
    expect(envelope.payloadHashAlg).toBe(CoseHashAlgorithm.SHA_256);
    expect(envelope.preimageContentType).toBe("text/plain");
  });

  test("can create hash envelope with payload location", async () => {
    const { createHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const data = new TextEncoder().encode("Test data");
    const envelope = await createHashEnvelope(data, {
      contentType: "application/json",
      location: "https://example.com/data.json",
    });

    expect(envelope.payloadLocation).toBe("https://example.com/data.json");
  });

  test("can stream hash from file", async () => {
    const { streamHashFromFile } = await import("../../../src/lib/cose/hash-envelope.ts");

    // Create test file
    const testFile = path.join(testDir, "test.bin");
    const testData = new Uint8Array(1024);
    for (let i = 0; i < testData.length; i++) {
      testData[i] = i % 256;
    }
    fs.writeFileSync(testFile, testData);

    const hash = await streamHashFromFile(testFile);

    expect(hash).toBeInstanceOf(Uint8Array);
    expect(hash.length).toBe(32); // SHA-256
  });

  test("can stream hash from large file", async () => {
    const { streamHashFromFile } = await import("../../../src/lib/cose/hash-envelope.ts");

    // Create 10MB test file
    const testFile = path.join(testDir, "large.bin");
    const chunkSize = 1024 * 1024; // 1MB chunks
    const file = fs.openSync(testFile, "w");

    for (let i = 0; i < 10; i++) {
      const chunk = new Uint8Array(chunkSize);
      for (let j = 0; j < chunkSize; j++) {
        chunk[j] = (i * chunkSize + j) % 256;
      }
      fs.writeSync(file, chunk);
    }
    fs.closeSync(file);

    const hash = await streamHashFromFile(testFile);

    expect(hash).toBeInstanceOf(Uint8Array);
    expect(hash.length).toBe(32);

    // Verify file size
    const stats = fs.statSync(testFile);
    expect(stats.size).toBe(10 * chunkSize);
  });

  test("streaming hash matches direct hash", async () => {
    const { streamHashFromFile, hashData } = await import("../../../src/lib/cose/hash-envelope.ts");

    // Create test file
    const testFile = path.join(testDir, "test.bin");
    const testData = new Uint8Array(10000);
    for (let i = 0; i < testData.length; i++) {
      testData[i] = i % 256;
    }
    fs.writeFileSync(testFile, testData);

    const streamHash = await streamHashFromFile(testFile);
    const directHash = await hashData(testData);

    expect(streamHash).toEqual(directHash);
  });

  test("can create envelope with custom hash algorithm", async () => {
    const { createHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const data = new TextEncoder().encode("Test");
    const envelope = await createHashEnvelope(data, {
      contentType: "text/plain",
      hashAlgorithm: CoseHashAlgorithm.SHA_256,
    });

    expect(envelope.payloadHashAlg).toBe(CoseHashAlgorithm.SHA_256);
  });

  test("handles empty data", async () => {
    const { createHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const data = new Uint8Array(0);
    const envelope = await createHashEnvelope(data, {
      contentType: "application/octet-stream",
    });

    expect(envelope.payloadHash).toBeInstanceOf(Uint8Array);
    expect(envelope.payloadHash.length).toBe(32);
  });
});

describe("Hash Envelope Validation", () => {
  test("can validate hash envelope", async () => {
    const { createHashEnvelope, validateHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const data = new TextEncoder().encode("Test data");
    const envelope = await createHashEnvelope(data, {
      contentType: "text/plain",
    });

    const isValid = await validateHashEnvelope(envelope, data);

    expect(isValid).toBe(true);
  });

  test("rejects tampered data", async () => {
    const { createHashEnvelope, validateHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const originalData = new TextEncoder().encode("Original data");
    const envelope = await createHashEnvelope(originalData, {
      contentType: "text/plain",
    });

    const tamperedData = new TextEncoder().encode("Tampered data");
    const isValid = await validateHashEnvelope(envelope, tamperedData);

    expect(isValid).toBe(false);
  });

  test("rejects tampered hash", async () => {
    const { createHashEnvelope, validateHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const data = new TextEncoder().encode("Test data");
    const envelope = await createHashEnvelope(data, {
      contentType: "text/plain",
    });

    // Tamper with hash
    envelope.payloadHash[0] ^= 0xFF;

    const isValid = await validateHashEnvelope(envelope, data);

    expect(isValid).toBe(false);
  });

  test("validates algorithm mismatch", async () => {
    const { createHashEnvelope, validateHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");

    const data = new TextEncoder().encode("Test data");
    const envelope = await createHashEnvelope(data, {
      contentType: "text/plain",
    });

    // Change algorithm
    envelope.payloadHashAlg = CoseHashAlgorithm.SHA_384;

    const isValid = await validateHashEnvelope(envelope, data);

    expect(isValid).toBe(false);
  });
});

describe("Hash Envelope with COSE Sign1", () => {
  test("can create signed hash envelope", async () => {
    const { generateES256KeyPair } = await import("../../../src/lib/cose/signer.ts");
    const { signHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");
    const { createCWTClaims } = await import("../../../src/lib/cose/sign.ts");

    const keyPair = await generateES256KeyPair();

    const cwtClaims = createCWTClaims({
      iss: "https://example.com/issuer",
    });

    const data = new TextEncoder().encode("Test data");
    const coseSign1 = await signHashEnvelope(
      data,
      {
        contentType: "text/plain",
        location: "https://example.com/data",
      },
      keyPair.privateKey,
      cwtClaims
    );

    expect(coseSign1.protected).toBeInstanceOf(Uint8Array);
    expect(coseSign1.payload).toBeInstanceOf(Uint8Array);
    expect(coseSign1.payload.length).toBe(32); // SHA-256 hash
    expect(coseSign1.signature).toBeInstanceOf(Uint8Array);
  });

  test("can verify signed hash envelope", async () => {
    const { generateES256KeyPair } = await import("../../../src/lib/cose/signer.ts");
    const { signHashEnvelope, verifyHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");
    const { createCWTClaims } = await import("../../../src/lib/cose/sign.ts");

    const keyPair = await generateES256KeyPair();

    const cwtClaims = createCWTClaims({
      iss: "https://example.com/issuer",
    });

    const data = new TextEncoder().encode("Test data");
    const coseSign1 = await signHashEnvelope(
      data,
      { contentType: "text/plain" },
      keyPair.privateKey,
      cwtClaims
    );

    const result = await verifyHashEnvelope(coseSign1, data, keyPair.publicKey);

    expect(result.signatureValid).toBe(true);
    expect(result.hashValid).toBe(true);
  });

  test("detects tampered artifact in signed envelope", async () => {
    const { generateES256KeyPair } = await import("../../../src/lib/cose/signer.ts");
    const { signHashEnvelope, verifyHashEnvelope } = await import("../../../src/lib/cose/hash-envelope.ts");
    const { createCWTClaims } = await import("../../../src/lib/cose/sign.ts");

    const keyPair = await generateES256KeyPair();

    const cwtClaims = createCWTClaims({
      iss: "https://example.com/issuer",
    });

    const originalData = new TextEncoder().encode("Original data");
    const coseSign1 = await signHashEnvelope(
      originalData,
      { contentType: "text/plain" },
      keyPair.privateKey,
      cwtClaims
    );

    const tamperedData = new TextEncoder().encode("Tampered data");
    const result = await verifyHashEnvelope(coseSign1, tamperedData, keyPair.publicKey);

    expect(result.signatureValid).toBe(true); // Signature is still valid
    expect(result.hashValid).toBe(false); // But hash doesn't match
  });

  test("can extract hash envelope parameters from protected header", async () => {
    const { generateES256KeyPair } = await import("../../../src/lib/cose/signer.ts");
    const { signHashEnvelope, extractHashEnvelopeParams } = await import("../../../src/lib/cose/hash-envelope.ts");
    const { createCWTClaims } = await import("../../../src/lib/cose/sign.ts");
    const { CoseHashAlgorithm } = await import("../../../src/lib/cose/constants.ts");

    const keyPair = await generateES256KeyPair();

    const cwtClaims = createCWTClaims({
      iss: "https://example.com/issuer",
    });

    const data = new TextEncoder().encode("Test data");
    const coseSign1 = await signHashEnvelope(
      data,
      {
        contentType: "application/json",
        location: "https://example.com/data.json",
      },
      keyPair.privateKey,
      cwtClaims
    );

    const params = extractHashEnvelopeParams(coseSign1);

    expect(params.payloadHashAlg).toBe(CoseHashAlgorithm.SHA_256);
    expect(params.preimageContentType).toBe("application/json");
    expect(params.payloadLocation).toBe("https://example.com/data.json");
  });
});

File: scitt-golang/pkg/cose/hash_envelope_test.go
package cose_test

import (
	"bytes"
	"os"
	"path/filepath"
	"testing"

	"github.com/tradeverifyd/transparency-service/scitt-golang/pkg/cose"
)

func TestCreateHashEnvelope(t *testing.T) {
	t.Run("creates hash envelope with default algorithm", func(t *testing.T) {
		data := []byte("Hello, World!")
		opts := cose.HashEnvelopeOptions{}

		envelope, err := cose.CreateHashEnvelope(data, opts)
		if err != nil {
			t.Fatalf("failed to create hash envelope: %v", err)
		}

		if envelope == nil {
			t.Fatal("envelope is nil")
		}
		if len(envelope.PayloadHash) == 0 {
			t.Error("payload hash is empty")
		}
		if envelope.PayloadHashAlg != cose.HashAlgorithmSHA256 {
			t.Errorf("expected SHA-256 algorithm (%d), got %d", cose.HashAlgorithmSHA256, envelope.PayloadHashAlg)
		}
	})

	t.Run("creates hash envelope with specified algorithm", func(t *testing.T) {
		data := []byte("Test data")
		opts := cose.HashEnvelopeOptions{
			HashAlgorithm: cose.HashAlgorithmSHA384,
		}

		envelope, err := cose.CreateHashEnvelope(data, opts)
		if err != nil {
			t.Fatalf("failed to create hash envelope: %v", err)
		}

		if envelope.PayloadHashAlg != cose.HashAlgorithmSHA384 {
			t.Errorf("expected SHA-384 algorithm, got %d", envelope.PayloadHashAlg)
		}
	})

	t.Run("creates hash envelope with metadata", func(t *testing.T) {
		data := []byte("Test data")
		opts := cose.HashEnvelopeOptions{
			ContentType: "application/json",
			Location:    "https://example.com/artifact.json",
		}

		envelope, err := cose.CreateHashEnvelope(data, opts)
		if err != nil {
			t.Fatalf("failed to create hash envelope: %v", err)
		}

		if envelope.PreimageContentType != opts.ContentType {
			t.Errorf("expected content type %s, got %s", opts.ContentType, envelope.PreimageContentType)
		}
		if envelope.PayloadLocation != opts.Location {
			t.Errorf("expected location %s, got %s", opts.Location, envelope.PayloadLocation)
		}
	})

	t.Run("produces consistent hashes", func(t *testing.T) {
		data := []byte("Consistent data")
		opts := cose.HashEnvelopeOptions{}

		envelope1, err := cose.CreateHashEnvelope(data, opts)
		if err != nil {
			t.Fatalf("failed to create envelope 1: %v", err)
		}

		envelope2, err := cose.CreateHashEnvelope(data, opts)
		if err != nil {
			t.Fatalf("failed to create envelope 2: %v", err)
		}

		if !bytes.Equal(envelope1.PayloadHash, envelope2.PayloadHash) {
			t.Error("hashes should be consistent for same data")
		}
	})

	t.Run("produces different hashes for different data", func(t *testing.T) {
		data1 := []byte("Data 1")
		data2 := []byte("Data 2")
		opts := cose.HashEnvelopeOptions{}

		envelope1, err := cose.CreateHashEnvelope(data1, opts)
		if err != nil {
			t.Fatalf("failed to create envelope 1: %v", err)
		}

		envelope2, err := cose.CreateHashEnvelope(data2, opts)
		if err != nil {
			t.Fatalf("failed to create envelope 2: %v", err)
		}

		if bytes.Equal(envelope1.PayloadHash, envelope2.PayloadHash) {
			t.Error("hashes should be different for different data")
		}
	})
}

func TestHashData(t *testing.T) {
	t.Run("hashes data with SHA-256", func(t *testing.T) {
		data := []byte("test")
		hash, err := cose.HashData(data, cose.HashAlgorithmSHA256)
		if err != nil {
			t.Fatalf("failed to hash data: %v", err)
		}

		if len(hash) != 32 {
			t.Errorf("expected 32 bytes for SHA-256, got %d", len(hash))
		}
	})

	t.Run("hashes data with SHA-384", func(t *testing.T) {
		data := []byte("test")
		hash, err := cose.HashData(data, cose.HashAlgorithmSHA384)
		if err != nil {
			t.Fatalf("failed to hash data: %v", err)
		}

		if len(hash) != 48 {
			t.Errorf("expected 48 bytes for SHA-384, got %d", len(hash))
		}
	})

	t.Run("hashes data with SHA-512", func(t *testing.T) {
		data := []byte("test")
		hash, err := cose.HashData(data, cose.HashAlgorithmSHA512)
		if err != nil {
			t.Fatalf("failed to hash data: %v", err)
		}

		if len(hash) != 64 {
			t.Errorf("expected 64 bytes for SHA-512, got %d", len(hash))
		}
	})

	t.Run("rejects unsupported algorithm", func(t *testing.T) {
		data := []byte("test")
		_, err := cose.HashData(data, 9999)
		if err == nil {
			t.Error("expected error for unsupported algorithm")
		}
	})
}

func TestStreamHashFromFile(t *testing.T) {
	// Create temporary directory
	tmpDir, err := os.MkdirTemp("", "cose-test-*")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	t.Run("hashes file content", func(t *testing.T) {
		// Create test file
		testFile := filepath.Join(tmpDir, "test.txt")
		testData := []byte("File content for hashing")
		if err := os.WriteFile(testFile, testData, 0644); err != nil {
			t.Fatalf("failed to write test file: %v", err)
		}

		// Hash file
		fileHash, err := cose.StreamHashFromFile(testFile, cose.HashAlgorithmSHA256)
		if err != nil {
			t.Fatalf("failed to hash file: %v", err)
		}

		// Hash data directly
		dataHash, err := cose.HashData(testData, cose.HashAlgorithmSHA256)
		if err != nil {
			t.Fatalf("failed to hash data: %v", err)
		}

		// Hashes should match
		if !bytes.Equal(fileHash, dataHash) {
			t.Error("file hash and data hash should match")
		}
	})

	t.Run("handles large files", func(t *testing.T) {
		// Create large test file (1MB)
		testFile := filepath.Join(tmpDir, "large.bin")
		largeData := make([]byte, 1024*1024)
		for i := range largeData {
			largeData[i] = byte(i % 256)
		}
		if err := os.WriteFile(testFile, largeData, 0644); err != nil {
			t.Fatalf("failed to write large file: %v", err)
		}

		// Hash file
		fileHash, err := cose.StreamHashFromFile(testFile, cose.HashAlgorithmSHA256)
		if err != nil {
			t.Fatalf("failed to hash large file: %v", err)
		}

		if len(fileHash) != 32 {
			t.Errorf("expected 32 bytes, got %d", len(fileHash))
		}
	})

	t.Run("rejects non-existent file", func(t *testing.T) {
		_, err := cose.StreamHashFromFile("/non/existent/file.txt", cose.HashAlgorithmSHA256)
		if err == nil {
			t.Error("expected error for non-existent file")
		}
	})
}

func TestValidateHashEnvelope(t *testing.T) {
	t.Run("validates correct hash", func(t *testing.T) {
		data := []byte("Test data")
		opts := cose.HashEnvelopeOptions{}

		envelope, err := cose.CreateHashEnvelope(data, opts)
		if err != nil {
			t.Fatalf("failed to create envelope: %v", err)
		}

		valid, err := cose.ValidateHashEnvelope(envelope, data)
		if err != nil {
			t.Fatalf("validation failed: %v", err)
		}

		if !valid {
			t.Error("hash should be valid")
		}
	})

	t.Run("rejects incorrect data", func(t *testing.T) {
		originalData := []byte("Original data")
		tamperedData := []byte("Tampered data")
		opts := cose.HashEnvelopeOptions{}

		envelope, err := cose.CreateHashEnvelope(originalData, opts)
		if err != nil {
			t.Fatalf("failed to create envelope: %v", err)
		}

		valid, err := cose.ValidateHashEnvelope(envelope, tamperedData)
		if err != nil {
			t.Fatalf("validation failed: %v", err)
		}

		if valid {
			t.Error("tampered data should not validate")
		}
	})
}

func TestSignHashEnvelope(t *testing.T) {
	// Generate test key pair
	keyPair, err := cose.GenerateES256KeyPair()
	if err != nil {
		t.Fatalf("failed to generate key pair: %v", err)
	}

	signer, err := cose.NewES256Signer(keyPair.Private)
	if err != nil {
		t.Fatalf("failed to create signer: %v", err)
	}

	t.Run("signs hash envelope", func(t *testing.T) {
		artifact := []byte("Artifact to sign")
		opts := cose.HashEnvelopeOptions{
			ContentType:   "application/octet-stream",
			Location:      "https://example.com/artifact.bin",
			HashAlgorithm: cose.HashAlgorithmSHA256,
		}

		coseSign1, err := cose.SignHashEnvelope(artifact, opts, signer, nil, false)
		if err != nil {
			t.Fatalf("failed to sign hash envelope: %v", err)
		}

		if coseSign1 == nil {
			t.Fatal("COSE Sign1 is nil")
		}
		if len(coseSign1.Signature) == 0 {
			t.Error("signature is empty")
		}
		if len(coseSign1.Payload) == 0 {
			t.Error("payload (hash) is empty")
		}
	})

	t.Run("signs with CWT claims", func(t *testing.T) {
		artifact := []byte("Artifact with claims")
		opts := cose.HashEnvelopeOptions{}
		cwtClaims := cose.CreateCWTClaims(cose.CWTClaimsOptions{
			Iss: "https://issuer.example.com",
			Sub: "artifact-123",
		})

		coseSign1, err := cose.SignHashEnvelope(artifact, opts, signer, cwtClaims, false)
		if err != nil {
			t.Fatalf("failed to sign with claims: %v", err)
		}

		// Verify claims are in protected headers
		headers, err := cose.GetProtectedHeaders(coseSign1)
		if err != nil {
			t.Fatalf("failed to get headers: %v", err)
		}

		if _, ok := headers[uint64(cose.HeaderLabelCWTClaims)]; !ok {
			t.Error("CWT claims should be in protected headers")
		}
	})

	t.Run("creates detached signature", func(t *testing.T) {
		artifact := []byte("Detached artifact")
		opts := cose.HashEnvelopeOptions{}

		coseSign1, err := cose.SignHashEnvelope(artifact, opts, signer, nil, true)
		if err != nil {
			t.Fatalf("failed to create detached signature: %v", err)
		}

		if coseSign1.Payload != nil {
			t.Error("detached signature should have nil payload")
		}
	})
}

func TestVerifyHashEnvelope(t *testing.T) {
	// Generate test key pair
	keyPair, err := cose.GenerateES256KeyPair()
	if err != nil {
		t.Fatalf("failed to generate key pair: %v", err)
	}

	signer, err := cose.NewES256Signer(keyPair.Private)
	if err != nil {
		t.Fatalf("failed to create signer: %v", err)
	}

	verifier, err := cose.NewES256Verifier(keyPair.Public)
	if err != nil {
		t.Fatalf("failed to create verifier: %v", err)
	}

	t.Run("verifies valid hash envelope", func(t *testing.T) {
		artifact := []byte("Valid artifact")
		opts := cose.HashEnvelopeOptions{}

		coseSign1, err := cose.SignHashEnvelope(artifact, opts, signer, nil, false)
		if err != nil {
			t.Fatalf("failed to sign: %v", err)
		}

		result, err := cose.VerifyHashEnvelope(coseSign1, artifact, verifier)
		if err != nil {
			t.Fatalf("verification failed: %v", err)
		}

		if !result.SignatureValid {
			t.Error("signature should be valid")
		}
		if !result.HashValid {
			t.Error("hash should be valid")
		}
	})

	t.Run("detects tampered artifact", func(t *testing.T) {
		originalArtifact := []byte("Original artifact")
		tamperedArtifact := []byte("Tampered artifact")
		opts := cose.HashEnvelopeOptions{}

		coseSign1, err := cose.SignHashEnvelope(originalArtifact, opts, signer, nil, false)
		if err != nil {
			t.Fatalf("failed to sign: %v", err)
		}

		result, err := cose.VerifyHashEnvelope(coseSign1, tamperedArtifact, verifier)
		if err != nil {
			t.Fatalf("verification failed: %v", err)
		}

		if !result.SignatureValid {
			t.Error("signature should still be valid")
		}
		if result.HashValid {
			t.Error("hash should NOT be valid for tampered artifact")
		}
	})

	t.Run("detects tampered signature", func(t *testing.T) {
		artifact := []byte("Test artifact")
		opts := cose.HashEnvelopeOptions{}

		coseSign1, err := cose.SignHashEnvelope(artifact, opts, signer, nil, false)
		if err != nil {
			t.Fatalf("failed to sign: %v", err)
		}

		// Tamper with signature
		coseSign1.Signature[0] ^= 0xFF

		result, err := cose.VerifyHashEnvelope(coseSign1, artifact, verifier)
		if err != nil {
			t.Fatalf("verification failed: %v", err)
		}

		if result.SignatureValid {
			t.Error("tampered signature should NOT be valid")
		}
	})
}

func TestExtractHashEnvelopeParams(t *testing.T) {
	keyPair, err := cose.GenerateES256KeyPair()
	if err != nil {
		t.Fatalf("failed to generate key pair: %v", err)
	}

	signer, err := cose.NewES256Signer(keyPair.Private)
	if err != nil {
		t.Fatalf("failed to create signer: %v", err)
	}

	t.Run("extracts parameters from COSE Sign1", func(t *testing.T) {
		artifact := []byte("Test artifact")
		opts := cose.HashEnvelopeOptions{
			ContentType:   "text/plain",
			Location:      "https://example.com/file.txt",
			HashAlgorithm: cose.HashAlgorithmSHA256,
		}

		coseSign1, err := cose.SignHashEnvelope(artifact, opts, signer, nil, false)
		if err != nil {
			t.Fatalf("failed to sign: %v", err)
		}

		params, err := cose.ExtractHashEnvelopeParams(coseSign1)
		if err != nil {
			t.Fatalf("failed to extract params: %v", err)
		}

		if params.PayloadHashAlg != opts.HashAlgorithm {
			t.Errorf("expected hash alg %d, got %d", opts.HashAlgorithm, params.PayloadHashAlg)
		}
		if params.PreimageContentType != opts.ContentType {
			t.Errorf("expected content type %s, got %s", opts.ContentType, params.PreimageContentType)
		}
		if params.PayloadLocation != opts.Location {
			t.Errorf("expected location %s, got %s", opts.Location, params.PayloadLocation)
		}
	})

	t.Run("handles missing optional parameters", func(t *testing.T) {
		artifact := []byte("Test artifact")
		opts := cose.HashEnvelopeOptions{
			HashAlgorithm: cose.HashAlgorithmSHA256,
		}

		coseSign1, err := cose.SignHashEnvelope(artifact, opts, signer, nil, false)
		if err != nil {
			t.Fatalf("failed to sign: %v", err)
		}

		params, err := cose.ExtractHashEnvelopeParams(coseSign1)
		if err != nil {
			t.Fatalf("failed to extract params: %v", err)
		}

		if params.PreimageContentType != "" {
			t.Error("content type should be empty")
		}
		if params.PayloadLocation != "" {
			t.Error("location should be empty")
		}
	})
}
